

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Single Cell Transcriptomics (PBMC3k) &#8212; singleCellHaystack  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=365ca57ee442770a23c6" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=365ca57ee442770a23c6" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=365ca57ee442770a23c6"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'tutorials/pbmc3k';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Spatial Transcriptomics" href="spatial.html" />
    <link rel="prev" title="Toy Example" href="toy.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">singleCellHaystack  documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="toy.html">Toy Example</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Single Cell Transcriptomics (PBMC3k)</a></li>
<li class="toctree-l2"><a class="reference internal" href="spatial.html">Spatial Transcriptomics</a></li>
<li class="toctree-l2"><a class="reference internal" href="trajectory.html">Trajectory Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="hoca.html">Human Single Cell Atlas Gene Expression</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../releases.html">Releases</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../releases/0.0.5.html">0.0.5 (2023-07-01)</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference.html">Reference</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.haystack.html">singleCellHaystack.haystack</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.cluster_genes.html">singleCellHaystack.cluster_genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.plot_rand_fit.html">singleCellHaystack.plot_rand_fit</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.plot_pval_rank.html">singleCellHaystack.plot_pval_rank</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.plot_pval_hist.html">singleCellHaystack.plot_pval_hist</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.plot_gene_clusters.html">singleCellHaystack.plot_gene_clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generated/singleCellHaystack.plot_compare_ranks.html">singleCellHaystack.plot_compare_ranks</a></li>
</ul>
</li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/ddiez/singleCellHaystack-py" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/tutorials/pbmc3k.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Single Cell Transcriptomics (PBMC3k)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Load-data">Load data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Filter-some-cells">Filter some cells</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Run-haystack">Run haystack</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Check-some-QC-plots">Check some QC plots</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualizing-the-results">Visualizing the results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Clustering-results-into-gene-modules">Clustering results into gene modules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Export-results">Export results</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="Single-Cell-Transcriptomics-(PBMC3k)">
<h1>Single Cell Transcriptomics (PBMC3k)<a class="headerlink" href="#Single-Cell-Transcriptomics-(PBMC3k)" title="Permalink to this heading">#</a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import scanpy as sc
import singleCellHaystack as hs

sc.set_figure_params(facecolor=&quot;white&quot;, dpi=90)
</pre></div>
</div>
</div>
<section id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Permalink to this heading">#</a></h2>
<p>We load the human PBMC data available from scanpy.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.datasets.pbmc3k_processed()
adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2638 × 1838
    obs: &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;louvain&#39;
    var: &#39;n_cells&#39;
    uns: &#39;draw_graph&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
    obsm: &#39;X_pca&#39;, &#39;X_tsne&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fr&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=&quot;louvain&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/diez/miniconda3/envs/singleCellHaystack/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_5_1.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_5_1.png" style="width: 517px; height: 337px;" />
</div>
</div>
</section>
<section id="Filter-some-cells">
<h2>Filter some cells<a class="headerlink" href="#Filter-some-cells" title="Permalink to this heading">#</a></h2>
<p>We remove Megakaryocytes since they are not interesting in this example and they will drive the top results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mega = adata.obs.louvain.isin([&quot;Megakaryocytes&quot;])
mega.value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
False    2623
True       15
Name: louvain, dtype: int64
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = adata[~mega, :]
adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
View of AnnData object with n_obs × n_vars = 2623 × 1838
    obs: &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;louvain&#39;
    var: &#39;n_cells&#39;
    uns: &#39;draw_graph&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
    obsm: &#39;X_pca&#39;, &#39;X_tsne&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fr&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
</section>
<section id="Run-haystack">
<h2>Run haystack<a class="headerlink" href="#Run-haystack" title="Permalink to this heading">#</a></h2>
<p>We run <code class="docutils literal notranslate"><span class="pre">haystack</span></code> using PCA coordinates and otherwise default parameters. Since the data has been scaled, we need to pass the raw data to avoid negative numbers. Also this will allow <code class="docutils literal notranslate"><span class="pre">haystack</span></code> to run on all the genes, not only those to calculate the PCA coordinates.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>res = hs.haystack(adata.raw.to_adata(), coord=&quot;pca&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; starting haystack ...
&gt; entering array method ...
&gt; scaling coordinates ...
&gt; calculating feature stds ...
&gt; removing 6 genes with zero variance ...
&gt; calculating grid points ...
&gt; calculating distance to cells ...
&gt; calculating densities ...
&gt; calculating Q dist ...
&gt; calculating KLD for 13708 features ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 13708/13708 [00:02&lt;00:00, 6130.26it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; calculating feature&#39;s CV ...
&gt; selecting genes to randomize ...
&gt; calculating randomized KLD ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 100/100 [00:02&lt;00:00, 40.36it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; calculating P values ...
&gt; done.
</pre></div></div>
</div>
</section>
<section id="Check-some-QC-plots">
<h2>Check some QC plots<a class="headerlink" href="#Check-some-QC-plots" title="Permalink to this heading">#</a></h2>
<p>We can examine some of the QC plots. First the randomization fits. These are used to calculate KLD from randomized expression levels for a subset of genes, in order to estimate the values to the entire gene set.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_rand_fit(res, &quot;mean&quot;)
hs.plot_rand_fit(res, &quot;sd&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_14_0.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_14_0.png" style="width: 358px; height: 358px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_14_1.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_14_1.png" style="width: 353px; height: 358px;" />
</div>
</div>
<p>The ranking of logpval and distribution of pval gives us some idea of how many significant genes we can detect.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_pval_rank(res)
hs.plot_pval_hist(res)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_16_0.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_16_0.png" style="width: 358px; height: 338px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_16_1.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_16_1.png" style="width: 358px; height: 316px;" />
</div>
</div>
</section>
<section id="Visualizing-the-results">
<h2>Visualizing the results<a class="headerlink" href="#Visualizing-the-results" title="Permalink to this heading">#</a></h2>
<p>A pandas DataFrame with the results can be obtained. By default the results are sorted by logpval_adj.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sum = res.top_features(n=10)
sum
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>KLD</th>
      <th>pval</th>
      <th>pval_adj</th>
      <th>logpval</th>
      <th>logpval_adj</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>9290</th>
      <td>GZMB</td>
      <td>0.004207</td>
      <td>2.671463e-67</td>
      <td>3.662041e-63</td>
      <td>-66.573251</td>
      <td>-62.436277</td>
    </tr>
    <tr>
      <th>3200</th>
      <td>FGFBP2</td>
      <td>0.003713</td>
      <td>2.990241e-63</td>
      <td>4.099023e-59</td>
      <td>-62.524294</td>
      <td>-58.387320</td>
    </tr>
    <tr>
      <th>7149</th>
      <td>PRF1</td>
      <td>0.002337</td>
      <td>2.624538e-57</td>
      <td>3.597716e-53</td>
      <td>-56.580947</td>
      <td>-52.443973</td>
    </tr>
    <tr>
      <th>12799</th>
      <td>CD79A</td>
      <td>0.002090</td>
      <td>1.626349e-56</td>
      <td>2.229399e-52</td>
      <td>-55.788786</td>
      <td>-51.651812</td>
    </tr>
    <tr>
      <th>1802</th>
      <td>GNLY</td>
      <td>0.002155</td>
      <td>3.416672e-56</td>
      <td>4.683574e-52</td>
      <td>-55.466397</td>
      <td>-51.329423</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>13172</th>
      <td>A1BG-AS1</td>
      <td>0.000098</td>
      <td>9.969061e-01</td>
      <td>1.000000e+00</td>
      <td>-0.001346</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>344</th>
      <td>SYNC</td>
      <td>0.000088</td>
      <td>9.974878e-01</td>
      <td>1.000000e+00</td>
      <td>-0.001092</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1731</th>
      <td>PCYOX1</td>
      <td>0.000065</td>
      <td>9.980668e-01</td>
      <td>1.000000e+00</td>
      <td>-0.000840</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>13247</th>
      <td>CRKL</td>
      <td>0.000038</td>
      <td>9.988439e-01</td>
      <td>1.000000e+00</td>
      <td>-0.000502</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>9925</th>
      <td>STOML1</td>
      <td>0.000099</td>
      <td>9.989249e-01</td>
      <td>1.000000e+00</td>
      <td>-0.000467</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
<p>13708 rows × 6 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=sum.gene.iloc[:4], ncols=2, cmap=&quot;Spectral_r&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_20_0.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_20_0.png" style="width: 737px; height: 654px;" />
</div>
</div>
</section>
<section id="Clustering-results-into-gene-modules">
<h2>Clustering results into gene modules<a class="headerlink" href="#Clustering-results-into-gene-modules" title="Permalink to this heading">#</a></h2>
<p>We can cluster the results using K-means. We need to pass the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object, the haystack results and the number of clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>gene_mods = hs.cluster_genes(adata.raw.to_adata(), res, n_clusters=4, random_state=1)
gene_mods.sort_values(&quot;cluster&quot;).groupby(&quot;cluster&quot;).head(3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>cluster</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GZMB</td>
      <td>0</td>
    </tr>
    <tr>
      <th>48</th>
      <td>CCL5</td>
      <td>0</td>
    </tr>
    <tr>
      <th>46</th>
      <td>CTSW</td>
      <td>0</td>
    </tr>
    <tr>
      <th>63</th>
      <td>CEBPD</td>
      <td>1</td>
    </tr>
    <tr>
      <th>64</th>
      <td>SPI1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>71</th>
      <td>NCF2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CD79A</td>
      <td>2</td>
    </tr>
    <tr>
      <th>96</th>
      <td>PKIG</td>
      <td>2</td>
    </tr>
    <tr>
      <th>11</th>
      <td>HLA-DQA1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>61</th>
      <td>CD1C</td>
      <td>3</td>
    </tr>
    <tr>
      <th>6</th>
      <td>FCER1A</td>
      <td>3</td>
    </tr>
    <tr>
      <th>20</th>
      <td>CLEC10A</td>
      <td>3</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_gene_clusters(adata.raw.to_adata(), gene_mods, basis=&quot;umap&quot;, ncols=2, figsize=[7, 6],   color_map=&quot;Spectral_r&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_24_0.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_24_0.png" style="width: 607px; height: 522px;" />
</div>
</div>
<p>Check some of the genes in <code class="docutils literal notranslate"><span class="pre">Module</span> <span class="pre">0</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mod0 = gene_mods.groupby(&quot;cluster&quot;).get_group(0)
mod0.head(4)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>cluster</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GZMB</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FGFBP2</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PRF1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>GNLY</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=mod0.gene.iloc[:4], cmap=&quot;Spectral_r&quot;, ncols=2)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_pbmc3k_27_0.png" class="no-scaled-link" src="../_images/tutorials_pbmc3k_27_0.png" style="width: 737px; height: 654px;" />
</div>
</div>
</section>
<section id="Export-results">
<h2>Export results<a class="headerlink" href="#Export-results" title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>#sum.to_csv(&quot;pbmc3k-results.tsv&quot;)
</pre></div>
</div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="toy.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Toy Example</p>
      </div>
    </a>
    <a class="right-next"
       href="spatial.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Spatial Transcriptomics</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Load-data">Load data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Filter-some-cells">Filter some cells</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Run-haystack">Run haystack</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Check-some-QC-plots">Check some QC plots</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualizing-the-results">Visualizing the results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Clustering-results-into-gene-modules">Clustering results into gene modules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Export-results">Export results</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=365ca57ee442770a23c6"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=365ca57ee442770a23c6"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>