<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Analysis of Human Cell Atlas of Fetal Gene Expression &mdash; singleCellHaystack  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Reference" href="../reference.html" />
    <link rel="prev" title="Application to very large datasets" href="../hoca.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            singleCellHaystack
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../articles.html">Articles</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../toy.html">Toy example</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pbmc3k.html">Application to scRNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../spatial.html">Application to spatial transcriptomics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../trajectory.html">Application to trajectory analysis</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../hoca.html">Application to very large datasets</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Analysis of Human Cell Atlas of Fetal Gene Expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-data">Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Filter-genes">Filter genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Preprocess">Preprocess</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Run-haystack">Run haystack</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">singleCellHaystack</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../articles.html">Articles</a></li>
          <li class="breadcrumb-item"><a href="../hoca.html">Application to very large datasets</a></li>
      <li class="breadcrumb-item active">Analysis of Human Cell Atlas of Fetal Gene Expression</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/ddiez/singleCellHaystack-py/blob/maindocs/articles/hoca.ipynb" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Analysis-of-Human-Cell-Atlas-of-Fetal-Gene-Expression">
<h1>Analysis of Human Cell Atlas of Fetal Gene Expression<a class="headerlink" href="#Analysis-of-Human-Cell-Atlas-of-Fetal-Gene-Expression" title="Permalink to this heading"></a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import scanpy as sc
import singleCellHaystack as hs

sc.settings.set_figure_params(dpi=150, facecolor=&#39;white&#39;)
</pre></div>
</div>
</div>
<p>Here we show how singleCellHaystack can handle very large datasets. For this we use the Human Cell Atlas of Fetal Gene Expression dataset with ~4 million cells.</p>
</section>
<section id="Load-data">
<h1>Load data<a class="headerlink" href="#Load-data" title="Permalink to this heading"></a></h1>
<p>Reading the loom file is very slow and memory costly. It is better to do this independently and export the data in h5ad format.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
#adata = sc.read_loom(&quot;data/GSE156793_S3_gene_count.loom&quot;)
adata = sc.read_h5ad(&quot;output/adata.h5ad&quot;)
adata
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 17.7 s, sys: 16.3 s, total: 34.1 s
Wall time: 36.6 s
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4062980 × 63561
    obs: &#39;All_reads&#39;, &#39;Assay&#39;, &#39;Batch&#39;, &#39;Development_day&#39;, &#39;Exon_reads&#39;, &#39;Experiment_batch&#39;, &#39;Fetus_id&#39;, &#39;Intron_reads&#39;, &#39;Main_cluster_name&#39;, &#39;Main_cluster_umap_1&#39;, &#39;Main_cluster_umap_2&#39;, &#39;Organ&#39;, &#39;Organ_cell_lineage&#39;, &#39;RT_group&#39;, &#39;Sex&#39;, &#39;Size_Factor&#39;, &#39;batch&#39;, &#39;obs_names&#39;, &#39;sample&#39;
    var: &#39;exon_intron&#39;, &#39;gene_id&#39;, &#39;gene_short_name&#39;, &#39;gene_type&#39;, &#39;index&#39;, &#39;var_names&#39;
    uns: &#39;Main_cluster_name_colors&#39;
    obsm: &#39;X_umap&#39;
</pre></div></div>
</div>
</section>
<section id="Filter-genes">
<h1>Filter genes<a class="headerlink" href="#Filter-genes" title="Permalink to this heading"></a></h1>
<p>We remove genes expressed in less than 100 cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.filter_genes(adata, min_cells=100)
adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4062980 × 35686
    obs: &#39;All_reads&#39;, &#39;Assay&#39;, &#39;Batch&#39;, &#39;Development_day&#39;, &#39;Exon_reads&#39;, &#39;Experiment_batch&#39;, &#39;Fetus_id&#39;, &#39;Intron_reads&#39;, &#39;Main_cluster_name&#39;, &#39;Main_cluster_umap_1&#39;, &#39;Main_cluster_umap_2&#39;, &#39;Organ&#39;, &#39;Organ_cell_lineage&#39;, &#39;RT_group&#39;, &#39;Sex&#39;, &#39;Size_Factor&#39;, &#39;batch&#39;, &#39;obs_names&#39;, &#39;sample&#39;
    var: &#39;exon_intron&#39;, &#39;gene_id&#39;, &#39;gene_short_name&#39;, &#39;gene_type&#39;, &#39;index&#39;, &#39;var_names&#39;, &#39;n_cells&#39;
    uns: &#39;Main_cluster_name_colors&#39;
    obsm: &#39;X_umap&#39;
</pre></div></div>
</div>
</section>
<section id="Preprocess">
<h1>Preprocess<a class="headerlink" href="#Preprocess" title="Permalink to this heading"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
sc.pp.normalize_total(adata, target_sum=1e4)
sc.pp.log1p(adata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 33.5 s, sys: 2.81 s, total: 36.3 s
Wall time: 36.3 s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)
sc.pl.highly_variable_genes(adata)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_11_0.png" src="../_images/articles_hoca_11_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 47 s, sys: 13.4 s, total: 1min
Wall time: 50.8 s
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.raw = adata
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = adata[:, adata.var.highly_variable]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
sc.pp.scale(adata, max_value=10)
sc.tl.pca(adata, svd_solver=&#39;arpack&#39;, n_comps=100)
sc.pl.pca_variance_ratio(adata, n_pcs=100, log=True)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/diez/miniconda3/envs/scanpy/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:843: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_14_1.png" src="../_images/articles_hoca_14_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 9h 24min 44s, sys: 15min 6s, total: 9h 39min 50s
Wall time: 25min 50s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
sc.pp.neighbors(adata, n_pcs=50)
sc.tl.umap(adata)
sc.pl.umap(adata)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/diez/miniconda3/envs/scanpy/lib/python3.10/site-packages/umap/spectral.py:342: UserWarning: Exited at iteration 20 with accuracies
[0.01355775 0.01357453 0.01446526]
not reaching the requested tolerance 1e-08.
Use iteration 21 instead with accuracy
0.013865846475387812.

  eigenvalues, eigenvectors = scipy.sparse.linalg.lobpcg(
/Users/diez/miniconda3/envs/scanpy/lib/python3.10/site-packages/umap/spectral.py:342: UserWarning: Exited postprocessing with accuracies
[0.01355775 0.01357453 0.01446526]
not reaching the requested tolerance 1e-08.
  eigenvalues, eigenvectors = scipy.sparse.linalg.lobpcg(
/Users/diez/miniconda3/envs/scanpy/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39;, &#39;norm&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_15_1.png" src="../_images/articles_hoca_15_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 2h 39min 8s, sys: 5min 54s, total: 2h 45min 2s
Wall time: 1h 45min 3s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
sc.tl.leiden(adata)
sc.pl.umap(adata, color=&quot;leiden&quot;, legend_loc=&quot;on data&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/diez/miniconda3/envs/scanpy/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_16_1.png" src="../_images/articles_hoca_16_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 6h 49min 54s, sys: 7min 32s, total: 6h 57min 27s
Wall time: 6h 56min 7s
</pre></div></div>
</div>
<p>We use the raw attribute so that we can test all genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_raw = adata.raw.to_adata()
adata_raw
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4062980 × 35686
    obs: &#39;All_reads&#39;, &#39;Assay&#39;, &#39;Batch&#39;, &#39;Development_day&#39;, &#39;Exon_reads&#39;, &#39;Experiment_batch&#39;, &#39;Fetus_id&#39;, &#39;Intron_reads&#39;, &#39;Main_cluster_name&#39;, &#39;Main_cluster_umap_1&#39;, &#39;Main_cluster_umap_2&#39;, &#39;Organ&#39;, &#39;Organ_cell_lineage&#39;, &#39;RT_group&#39;, &#39;Sex&#39;, &#39;Size_Factor&#39;, &#39;batch&#39;, &#39;obs_names&#39;, &#39;sample&#39;, &#39;leiden&#39;
    var: &#39;exon_intron&#39;, &#39;gene_id&#39;, &#39;gene_short_name&#39;, &#39;gene_type&#39;, &#39;index&#39;, &#39;var_names&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;Main_cluster_name_colors&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;
    obsm: &#39;X_umap&#39;, &#39;X_pca&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<p>We set the gene symbols as index, so that they will be included in the haystack results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_raw.var[&quot;index&quot;]=list(map(str, adata_raw.var.gene_short_name.to_list()))
adata_raw.var=adata_raw.var.set_index(&quot;index&quot;)
adata_raw.var
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>exon_intron</th>
      <th>gene_id</th>
      <th>gene_short_name</th>
      <th>gene_type</th>
      <th>var_names</th>
      <th>n_cells</th>
      <th>highly_variable</th>
      <th>means</th>
      <th>dispersions</th>
      <th>dispersions_norm</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>WASH7P</th>
      <td>exon</td>
      <td>ENSG00000227232.4</td>
      <td>WASH7P</td>
      <td>pseudogene</td>
      <td>ENSG00000227232.4</td>
      <td>1719</td>
      <td>False</td>
      <td>0.005195</td>
      <td>3.152367</td>
      <td>0.433195</td>
    </tr>
    <tr>
      <th>RP11-34P13.7</th>
      <td>exon</td>
      <td>ENSG00000238009.2</td>
      <td>RP11-34P13.7</td>
      <td>lincRNA</td>
      <td>ENSG00000238009.2</td>
      <td>2492</td>
      <td>False</td>
      <td>0.005935</td>
      <td>2.904171</td>
      <td>-0.624707</td>
    </tr>
    <tr>
      <th>CICP27</th>
      <td>exon</td>
      <td>ENSG00000233750.3</td>
      <td>CICP27</td>
      <td>pseudogene</td>
      <td>ENSG00000233750.3</td>
      <td>476</td>
      <td>False</td>
      <td>0.001367</td>
      <td>2.950321</td>
      <td>-0.427997</td>
    </tr>
    <tr>
      <th>AL627309.1</th>
      <td>exon</td>
      <td>ENSG00000237683.5</td>
      <td>AL627309.1</td>
      <td>protein_coding</td>
      <td>ENSG00000237683.5</td>
      <td>4140</td>
      <td>True</td>
      <td>0.016144</td>
      <td>3.347481</td>
      <td>1.264840</td>
    </tr>
    <tr>
      <th>RP11-34P13.15</th>
      <td>exon</td>
      <td>ENSG00000268903.1</td>
      <td>RP11-34P13.15</td>
      <td>pseudogene</td>
      <td>ENSG00000268903.1</td>
      <td>197</td>
      <td>False</td>
      <td>0.000877</td>
      <td>3.482546</td>
      <td>1.840537</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>RNA28S5</th>
      <td>exon</td>
      <td>ENSG00000266658.2</td>
      <td>RNA28S5</td>
      <td>rRNA</td>
      <td>ENSG00000266658.2</td>
      <td>902875</td>
      <td>False</td>
      <td>1.680279</td>
      <td>3.473712</td>
      <td>0.144642</td>
    </tr>
    <tr>
      <th>AL592170.1</th>
      <td>exon</td>
      <td>ENSG00000233094.2</td>
      <td>AL592170.1</td>
      <td>protein_coding</td>
      <td>ENSG00000233094.2</td>
      <td>131</td>
      <td>False</td>
      <td>0.000289</td>
      <td>2.926089</td>
      <td>-0.531284</td>
    </tr>
    <tr>
      <th>ZNF84</th>
      <td>exon</td>
      <td>ENSG00000218497.3</td>
      <td>ZNF84</td>
      <td>protein_coding</td>
      <td>ENSG00000218497.3</td>
      <td>147</td>
      <td>False</td>
      <td>0.000392</td>
      <td>3.225180</td>
      <td>0.743549</td>
    </tr>
    <tr>
      <th>AL603926.1</th>
      <td>exon</td>
      <td>ENSG00000229631.2</td>
      <td>AL603926.1</td>
      <td>protein_coding</td>
      <td>ENSG00000229631.2</td>
      <td>4936</td>
      <td>False</td>
      <td>0.012115</td>
      <td>2.908165</td>
      <td>-0.607682</td>
    </tr>
    <tr>
      <th>CU459201.1</th>
      <td>exon</td>
      <td>ENSG00000251180.2</td>
      <td>CU459201.1</td>
      <td>protein_coding</td>
      <td>ENSG00000251180.2</td>
      <td>163</td>
      <td>False</td>
      <td>0.000411</td>
      <td>2.841433</td>
      <td>-0.892118</td>
    </tr>
  </tbody>
</table>
<p>35686 rows × 10 columns</p>
</div></div>
</div>
</section>
<section id="Run-haystack">
<h1>Run haystack<a class="headerlink" href="#Run-haystack" title="Permalink to this heading"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%time
res = hs.haystack(adata_raw, &quot;pca&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; starting haystack ...
&gt; entering array method ...
&gt; scaling coordinates ...
&gt; calculating feature stds ...
&gt; calculating grid points ...
&gt; calculating distance to cells ...
&gt; calculating densities ...
&gt; calculating Q dist ...
&gt; calculating KLD for 35686 features ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 35686/35686 [41:49&lt;00:00, 14.22it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; calculating feature&#39;s CV ...
&gt; selecting genes to randomize ...
&gt; calculating randomized KLD ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [1:26:16&lt;00:00, 51.77s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; calculating P values ...
&gt; done.
CPU times: user 15h 7min 37s, sys: 1d 2h 37min 17s, total: 1d 17h 44min 55s
Wall time: 3h 38min 14s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_rand_fit(res, &quot;mean&quot;)
hs.plot_rand_fit(res, &quot;sd&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_23_0.png" src="../_images/articles_hoca_23_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_23_1.png" src="../_images/articles_hoca_23_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_pval_rank(res)
hs.plot_pval_hist(res)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_24_0.png" src="../_images/articles_hoca_24_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_24_1.png" src="../_images/articles_hoca_24_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sum = res[&quot;results&quot;]
sum.head(10)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>KLD</th>
      <th>pval</th>
      <th>pval_adj</th>
      <th>logpval</th>
      <th>logpval_adj</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>19880</th>
      <td>MALAT1</td>
      <td>0.000011</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-8204.908347</td>
      <td>-8200.355849</td>
    </tr>
    <tr>
      <th>16291</th>
      <td>PTPRD</td>
      <td>0.000123</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-8133.804309</td>
      <td>-8129.251811</td>
    </tr>
    <tr>
      <th>13709</th>
      <td>AUTS2</td>
      <td>0.000050</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-7864.983675</td>
      <td>-7860.431177</td>
    </tr>
    <tr>
      <th>3996</th>
      <td>NRXN1</td>
      <td>0.000099</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-7552.591193</td>
      <td>-7548.038695</td>
    </tr>
    <tr>
      <th>6733</th>
      <td>ROBO2</td>
      <td>0.000128</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-7056.877873</td>
      <td>-7052.325375</td>
    </tr>
    <tr>
      <th>12169</th>
      <td>BAI3</td>
      <td>0.000103</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-6534.973548</td>
      <td>-6530.421050</td>
    </tr>
    <tr>
      <th>8452</th>
      <td>LPHN3</td>
      <td>0.000097</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-5932.948257</td>
      <td>-5928.395759</td>
    </tr>
    <tr>
      <th>6943</th>
      <td>LSAMP</td>
      <td>0.000162</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-5924.083997</td>
      <td>-5919.531500</td>
    </tr>
    <tr>
      <th>29615</th>
      <td>DLGAP1</td>
      <td>0.000110</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-5834.298885</td>
      <td>-5829.746387</td>
    </tr>
    <tr>
      <th>22249</th>
      <td>ANKS1B</td>
      <td>0.000108</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>-5747.931987</td>
      <td>-5743.379489</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>genes = sum.head(11).gene.tolist()
genes
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;MALAT1&#39;,
 &#39;PTPRD&#39;,
 &#39;AUTS2&#39;,
 &#39;NRXN1&#39;,
 &#39;ROBO2&#39;,
 &#39;BAI3&#39;,
 &#39;LPHN3&#39;,
 &#39;LSAMP&#39;,
 &#39;DLGAP1&#39;,
 &#39;ANKS1B&#39;,
 &#39;SYT1&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata_raw, color=[&quot;leiden&quot;] + genes, color_map=&quot;Spectral_r&quot;, legend_loc=None, sort_order=False)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/diez/miniconda3/envs/scanpy/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_hoca_27_1.png" src="../_images/articles_hoca_27_1.png" />
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../hoca.html" class="btn btn-neutral float-left" title="Application to very large datasets" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../reference.html" class="btn btn-neutral float-right" title="Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>