<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Load data &mdash; singleCellHaystack  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Application to very large datasets" href="../hoca.html" />
    <link rel="prev" title="Application to trajectory analysis" href="../trajectory.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            singleCellHaystack
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../articles.html">Articles</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../toy.html">Toy example</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pbmc3k.html">Application to scRNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../spatial.html">Application to spatial transcriptomics</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../trajectory.html">Application to trajectory analysis</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Preprocess">Preprocess</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Calculate-DPT-trajectories">Calculate DPT trajectories</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Run-singleCellHaystack">Run singleCellHaystack</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Export-results">Export results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../hoca.html">Application to very large datasets</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">singleCellHaystack</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../articles.html">Articles</a></li>
          <li class="breadcrumb-item"><a href="../trajectory.html">Application to trajectory analysis</a></li>
      <li class="breadcrumb-item active">Load data</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/ddiez/singleCellHaystack-py/blob/maindocs/articles/trajectory.ipynb" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import scanpy as sc
import singleCellHaystack as hs
</pre></div>
</div>
</div>
<section id="Load-data">
<h1>Load data<a class="headerlink" href="#Load-data" title="Permalink to this heading"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.datasets.paul15()
adata
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: In Scanpy 0.*, this returned logarithmized data. Now it returns non-logarithmized data.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/scanpy/datasets/_datasets.py:202: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.
  adata.uns[&#39;iroot&#39;] = 840
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2730 × 3451
    obs: &#39;paul15_clusters&#39;
    uns: &#39;iroot&#39;
</pre></div></div>
</div>
</section>
<section id="Preprocess">
<h1>Preprocess<a class="headerlink" href="#Preprocess" title="Permalink to this heading"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.normalize_total(adata)
sc.pp.log1p(adata)
sc.tl.pca(adata)
sc.pl.pca(adata)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39;, &#39;norm&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_4_1.png" src="../_images/articles_trajectory_4_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata)
sc.tl.leiden(adata)
sc.tl.umap(adata)
sc.pl.umap(adata, color=&quot;leiden&quot;, legend_loc=&quot;on data&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/home/diez/miniconda3/envs/singleCellHaystack/lib/python3.11/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_5_1.png" src="../_images/articles_trajectory_5_1.png" />
</div>
</div>
</section>
<section id="Calculate-DPT-trajectories">
<h1>Calculate DPT trajectories<a class="headerlink" href="#Calculate-DPT-trajectories" title="Permalink to this heading"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.tl.diffmap(adata)
sc.tl.dpt(adata)
sc.pl.umap(adata, color=&quot;dpt_pseudotime&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_7_0.png" src="../_images/articles_trajectory_7_0.png" />
</div>
</div>
</section>
<section id="Run-singleCellHaystack">
<h1>Run singleCellHaystack<a class="headerlink" href="#Run-singleCellHaystack" title="Permalink to this heading"></a></h1>
<p>For now there is not direct way to access the pseudotime information from AnnData object. But we can pass the expression matrix and 1D coordinates directly.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import numpy
pseudotime = numpy.array(adata.obs.dpt_pseudotime.to_list())
pseudotime = pseudotime.reshape(-1, 1)
pseudotime
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[0.0275316965],
       [0.1844685227],
       [0.2824616432],
       ...,
       [0.0244765058],
       [0.1840221733],
       [0.2610901594]])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>res = hs.haystack(adata.X, coord=pseudotime, features=adata.var_names)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; entering array method ...
&gt; scaling coordinates ...
&gt; calculating feature stds ...
&gt; calculating grid points ...
&gt; calculating distance to cells ...
&gt; calculating densities ...
&gt; calculating Q dist ...
&gt; calculating KLD for 3451 features ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|█████████████████████████████████████████████████████████████████████████████| 3451/3451 [00:01&lt;00:00, 2068.14it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; calculating feature&#39;s CV ...
&gt; selecting genes to randomize ...
&gt; calculating randomized KLD ...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|█████████████████████████████████████████████████████████████████████████████████| 100/100 [00:04&lt;00:00, 20.31it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&gt; calculating P values ...
&gt; done.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_rand_fit(res, &quot;mean&quot;)
hs.plot_rand_fit(res, &quot;sd&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_12_0.png" src="../_images/articles_trajectory_12_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_12_1.png" src="../_images/articles_trajectory_12_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>hs.plot_pval_rank(res)
hs.plot_pval_hist(res)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_13_0.png" src="../_images/articles_trajectory_13_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_13_1.png" src="../_images/articles_trajectory_13_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sum = res.top_features(9)
sum
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>KLD</th>
      <th>CV</th>
      <th>pval</th>
      <th>pval_adj</th>
      <th>logpval</th>
      <th>logpval_adj</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1121</th>
      <td>Fam132a</td>
      <td>0.354324</td>
      <td>1.111779</td>
      <td>1.907912e-81</td>
      <td>6.584203e-78</td>
      <td>-80.719442</td>
      <td>-77.181497</td>
    </tr>
    <tr>
      <th>1954</th>
      <td>Mt2</td>
      <td>0.476056</td>
      <td>1.326631</td>
      <td>1.375929e-77</td>
      <td>4.748330e-74</td>
      <td>-76.861404</td>
      <td>-73.323459</td>
    </tr>
    <tr>
      <th>2590</th>
      <td>Rhd</td>
      <td>0.601656</td>
      <td>1.491535</td>
      <td>3.467935e-77</td>
      <td>1.196784e-73</td>
      <td>-76.459929</td>
      <td>-72.921984</td>
    </tr>
    <tr>
      <th>2425</th>
      <td>Prtn3</td>
      <td>0.164153</td>
      <td>0.776379</td>
      <td>1.141090e-74</td>
      <td>3.937903e-71</td>
      <td>-73.942680</td>
      <td>-70.404735</td>
    </tr>
    <tr>
      <th>1040</th>
      <td>Elane</td>
      <td>0.291931</td>
      <td>1.066940</td>
      <td>1.240241e-74</td>
      <td>4.280072e-71</td>
      <td>-73.906494</td>
      <td>-70.368549</td>
    </tr>
    <tr>
      <th>2910</th>
      <td>Sphk1</td>
      <td>0.568372</td>
      <td>1.485415</td>
      <td>3.270921e-74</td>
      <td>1.128795e-70</td>
      <td>-73.485330</td>
      <td>-69.947385</td>
    </tr>
    <tr>
      <th>815</th>
      <td>Ctsg</td>
      <td>0.212397</td>
      <td>0.907998</td>
      <td>8.823847e-74</td>
      <td>3.045110e-70</td>
      <td>-73.054342</td>
      <td>-69.516397</td>
    </tr>
    <tr>
      <th>1072</th>
      <td>Ermap</td>
      <td>0.440583</td>
      <td>1.333001</td>
      <td>1.493464e-72</td>
      <td>5.153945e-69</td>
      <td>-71.825805</td>
      <td>-68.287860</td>
    </tr>
    <tr>
      <th>778</th>
      <td>Cpox</td>
      <td>0.349522</td>
      <td>1.195660</td>
      <td>6.859488e-72</td>
      <td>2.367209e-68</td>
      <td>-71.163708</td>
      <td>-67.625763</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=sum.gene, ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/articles_trajectory_15_0.png" src="../_images/articles_trajectory_15_0.png" />
</div>
</div>
</section>
<section id="Export-results">
<h1>Export results<a class="headerlink" href="#Export-results" title="Permalink to this heading"></a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>#sum.to_csv(&quot;results-trajectory.csv&quot;)
</pre></div>
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../trajectory.html" class="btn btn-neutral float-left" title="Application to trajectory analysis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../hoca.html" class="btn btn-neutral float-right" title="Application to very large datasets" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>